VolumeWidget/ArticleWidget/ArticleWidgetProvider.swift

```diff
@@ -0,0 +1,134 @@
+//
+//  ArticleWidgetProvider.swift
+//  Volume
+//
+//  Created by Vin Bui on 11/11/23.
+//  Copyright © 2023 Cornell AppDev. All rights reserved.
+//
+
+import Combine
+import WidgetKit
+
+struct ArticleWidgetProvider: TimelineProvider {
+
+    // MARK: - Properties
+
+    private let articlesLimit: Double = 4
+
+    class ProviderCancellable {
+        static var queryBag = Set<AnyCancellable>()
+    }
+
+    // MARK: - TimelineProvider Methods
+
+    /// Provides an Article entry representing a placeholder version of the Article widget.
+    func placeholder(in context: Context) -> ArticleEntry {
+        ArticleEntry(date: .now, article: ArticleWidgetProvider.dummyArticle!)
+    }
+
+    /// Provides an Article entry representing the current time and state of the Article widget.
+    func getSnapshot(in context: Context, completion: @escaping (ArticleEntry) -> Void) {
+        let entry = ArticleEntry(date: .now, article: ArticleWidgetProvider.dummyArticle!)
+        completion(entry)
+    }
+
+    /// Provides an array of Article entries for the current time and any future times to update the Article widget.
+    func getTimeline(in context: Context, completion: @escaping (Timeline<ArticleEntry>) -> Void) {
+        fetchTrendingArticles { articles in
+            var entries: [ArticleEntry] = []
+
+            // Generate a timeline with at most 4 entries every hour
+            for hourOffset in 0..<articles.count {
+                let entryDate = Calendar.current.date(byAdding: .hour, value: hourOffset, to: .now)
+                let entry = ArticleEntry(date: entryDate!, article: articles[hourOffset])
+                entries.append(entry)
+            }
+
+            let timeline = Timeline(entries: entries, policy: .atEnd)
+            completion(timeline)
+        }
+    }
+
+    // MARK: - Network Requests
+
+    private func fetchTrendingArticles(completion: @escaping ([Article]) -> Void) {
```

Comment:
This should be coming from some shared NetworkManager

Reply:
It's sloppy to include networking code inside the Provider. Separate the business logic.

Reply:
Apollo auto generates an `API.swift` file which is the equivalent to a `NetworkManager` file. Our `Network.swift` class allows us to use it with Combine by creating publishers and subscribers. But yea you're right this should be filtered out to a ViewModel.